rocketry/tasks/func.py

```diff
@@ -173,7 +179,7 @@ def __init__(self, func=None, **kwargs):
             # We initiate the class lazily by creating
             # almost empty shell class that is populated
             # in next __call__ (which should occur immediately)
-            self._delayed_kwargs = kwargs
+            FuncTask._delayed_kwargs = kwargs
```

Comment:
The same applies here. I think these should be instance attributes. It probably works this way but only if one does nothing out of the ordinary before setting the func (which is surprisingly common for users).

Reply:
This is a tricky one, I hated this solution but it was the only way I could get this to work.

For some background, I'm not the most knowledgable when it comes to pydantic so I'm not completely sure how this use to be, but one of the most recent migrations is this to Model config:

> To instead preserve any extra data provided, you can set extra='allow'. The extra fields will then be stored in `BaseModel.__pydantic_extra__`

extra="allow" has been set in the Task config to allow for this. But its mainly installing attributes that are unknown in `__pydantic_extra__`, I'm not sure if this functionality is new or not.

If I set those values back to self I get the following error:
```
Exception has occurred: AttributeError
'FuncTask' object has no attribute '__pydantic_extra__'
  File "/home/jypo/rocketry-pydanticv2/rocketry/rocketry/tasks/func.py", line 182, in __init__
    self._delayed_kwargs = kwargs
  File "/home/jypo/rocketry-pydanticv2/rocketry/rocketry/session.py", line 420, in create_task
    return FuncTask(name_include_module=False, _name_template='{func_name}', session=self, **kwargs)
  File "/home/jypo/rocketry-pydanticv2/rocketry/rocketry/application.py", line 20, in task
    return self.session.create_task(start_cond=start_cond, name=name, **kwargs)
  File "/home/jypo/rocketry-pydanticv2/rocketry/test.py", line 28, in <module>
    @app.task("every 3 seconds")
AttributeError: 'FuncTask' object has no attribute '__pydantic_extra__'
```

The odd part here is that _delay_kwargs is set in the class and it also has a default value, so why its trying to treat this as an "extra" variable I have no idea.

Additionally, I'm not completely sure if its the order of execution of the code, but when I run a debugger to try and see if `__pydantic_extra__` does exist in the class it seems to have a traceback at this point.

![image](https://github.com/Miksus/rocketry/assets/77198432/bca63bad-6fad-408f-a37f-5f719f740155)

The only way around this is to store delayed_kwargs inside the un-initated class variable (which is a bit shit).

I'll try and take another look at this one and see if I can come up with a better solution.

Reply:
So this has taken a bit of reverse engineering on the pydantic part. But this issue is to do with pydantic some kind of order of execution or something along those lines.

So how I best understand it. When we use the `@app.task` decorator it returns a FuncTask
FuncTask inherits from Task which inherits from BaseModel (FuncTask > Task > BaseModel).

Because both FuncTask and Task have overridden `__init__` for the pydantic BaseModel (until `super().__init__()` is run) it doesn't fully initiate (which is why we see tracebacks in the model) while its executing FuncTasks `__init__`

The variables do partially exist (I'm presuming because of the validation being run before the init partially initiating a pydantic BaseModel?). So essentially, it looks like its trying to use pydantic functionality without it being fully initated (I think)

### Horrible workaround
I tried calling model_construct to try and rebuild self using the BaseModel init but most solutions I tried broke the class in some form by overwriting existing variables

So I had a look at the pydantic code and mainly the way that the ["model construct"](https://github.com/pydantic/pydantic/blob/2181d12829f008431fead7956c909dc88d140725/pydantic/main.py#L193) function worked.

It has set a global variable (of a function) inside main `_object_setattr` and during the construct it calls the two following lines:
```
#inside a couple if statements
_object_setattr(m, '__pydantic_private__', None)
_object_setattr(m, '__pydantic_extra__', _extra)
```

I imported this functionality and set those manually within the FuncTask `__init__ ` function and it seems to work. But its perhaps an awful solution and clearly not intended to be used.

```
from pydantic.main import _object_setattr
...
...
class FuncTask(Task):
...
...
    def __init__(self, func=None, **kwargs):
        only_func_set = func is not None and not kwargs
        no_func_set = func is None and kwargs.get('path') is None
        _object_setattr(self, "__pydantic_extra__", {})
        _object_setattr(self, "__pydantic_private__", None)
        if no_func_set:
            # FuncTask was probably called like:
            # @FuncTask(...)
            # def myfunc(...): ...

            # We initiate the class lazily by creating
            # almost empty shell class that is populated
            # in next __call__ (which should occur immediately)
            self._delayed_kwargs = kwargs
            return
        if only_func_set:
...
```

This isn't a very elegant solution, but it looks like the best that can be done without involving a massive rework of some of this functionality with delayed_kwargs?

What would be your opinion on this? I'll commit these changes for now, but can easily revert if needed.